### DISCUSSION TITLE
Weird result using queries' union_all()

### DISCUSSION THREAD

**Discussion by sbq-clements** [sqlalchemy/sqlalchemy#9101] [Usage Questions] [answered]
Hi! 
Here's a bit of code (not updated for the 2.0 style) that does not gives the same result depending on if I use SqlAlchemy 1.3.23 or 1.4.46:

```python
from sqlalchemy import (
    Column,
    create_engine,
    ForeignKey,
    Integer,
    String,
)

from sqlalchemy.ext.declarative import declarative_base

from sqlalchemy.orm import (
    query_expression,
    relationship,
    with_expression,
    Session,
)


engine = create_engine("sqlite:///:memory:")
session = Session(engine)
Base = declarative_base(bind=engine)

class Author(Base):
    __tablename__ = "authors"

    id = Column(Integer, primary_key=True)
    books = relationship(
        "Book",
        cascade="all",
        lazy="dynamic",
    )


class Book(Base):
    __tablename__ = "books"

    id = Column(Integer, primary_key=True)
    expr = query_expression()

    author_id = Column(
        ForeignKey("authors.id"),
        nullable=False,
    )
    author = relationship(Author, back_populates="books")


def gen_query(a):
    query = a.books.options(with_expression(Book.expr, Book.id + a.id))
    return query


Base.metadata.create_all(engine)

author1 = Author(id=1)
author2 = Author(id=2)
authors = [author1, author2]

session.add(Book(author=author2))
session.add(Book(author=author2))
for author in authors:
    session.add(author)
session.commit()

queries = [gen_query(author) for author in authors]
query = queries[0].union_all(*queries[1:])

for book in query:
    print(f"Book {book.id}: expr = {book.expr}, should be {book.id+book.author_id}")
```

Result in 1.4.46 (weird?):
```
Book 1: expr = 2, should be 3
Book 2: expr = 3, should be 4
```

Result in 1.3.23 (correct?):
```
Book 1: expr = 3, should be 3
Book 2: expr = 4, should be 4
```

If I understood correctly, one of the reason queries are being entirely replaced by selects is the fact that queries `query.union()` doesn't work well, but the fact that this code gives a different result depending on the library's version seems problematic to me?
Thanks!

**Comment by zzzeek** [comment, marked as answer]
this is expected behavior but unfortunately I cannot locate any documentation of this change, nor is there a note clarifying the purpose of with_expression().   this is probably the same issue as #8881, looks like the same query.

with_expression() is an ORM loader option.  It is only intended to affect data that is applied to an ORM object being returned, so therefore is not applicable to subqueries or queries that are looking for specific sets of columns.

the pattern shown at https://github.com/sqlalchemy/sqlalchemy/issues/8881#issuecomment-1327874263 would be the idiomatic way to get this pattern.

```py
def gen_query(a):
    query = a.books.add_columns((Book.id + a.id).label("my_expr"))
    return query

queries = [gen_query(author) for author in authors]

subq = union_all(*queries)
stmt = (
    select(Book)
    .from_statement(subq)
    .options(with_expression(Book.expr, subq.selected_columns.my_expr))
)

```

you'll note the generated SQL doesnt use a subquery.

**Reply by sbq-clements** [reply]
I know about this way to write the code (because I had already found https://github.com/sqlalchemy/sqlalchemy/issues/8881).

I was expecting for example to *lose* the value of the expression like in this issue but I got worried because this of code behaved differently between two versions of the code, and becaus the difference was not that the expression value was lost but that it was actually wrong.

Thanks!

**Reply by zzzeek** [reply]
the query.union_all() situation has always been terrible so in designing the SQLAlchemy 2.0 approach (where most of it is present in 1.4) I simply had to let its use cases go to some extent.

I've written a new upgrade guide for this issue which is at https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#column-loaders-such-as-deferred-with-expression-only-take-effect-when-indicated-on-the-outermost-full-entity-query
